src/AzureDeprecation.APIs.REST/DataAccess/Repositories/CosmosDbDeprecationsRepository.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System.Net;
+using System.Runtime.CompilerServices;
+using AzureDeprecation.APIs.REST.DataAccess.Interfaces;
+using AzureDeprecation.APIs.REST.DataAccess.Models;
+using AzureDeprecation.APIs.REST.DataAccess.Utils;
+using AzureDeprecation.APIs.REST.Settings;
+using CodeJam;
+using Microsoft.Azure.Cosmos;
+using Microsoft.Azure.Cosmos.Linq;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AzureDeprecation.APIs.REST.DataAccess.Repositories;
+
+internal class CosmosDbDeprecationsRepository : IDeprecationsRepository
+{
+    readonly CosmosClient _client;
+    readonly CosmosDbOptions _dbOptions;
+    readonly ILogger<CosmosDbDeprecationsRepository> _logger;
+
+    bool _isDatabaseInitialized;
+    readonly SemaphoreSlim _locker = new(1, 1);
+
+    public CosmosDbDeprecationsRepository(
+        CosmosClient cosmosClient,
+        IOptions<CosmosDbOptions> dbSettings,
+        ILogger<CosmosDbDeprecationsRepository> logger)
+    {
+        Code.NotNull(dbSettings?.Value, nameof(dbSettings));
+        Code.NotNullNorEmpty(dbSettings.Value.ConnectionString, nameof(dbSettings.Value.ConnectionString));
+
+        _client = cosmosClient;
+        _dbOptions = dbSettings.Value;
+        _logger = logger;
+    }
+
+    public async IAsyncEnumerable<NoticeEntity> GetDeprecationsAsync(
+        DeprecationsRequestModel deprecationsRequestModel,
+        [EnumeratorCancellation] CancellationToken cancellation = default)
+    {
+        await EnsureDatabaseInitializedAsync(cancellation);
```

Comment:
I think we're overkilling it here, did you check the Cosmos SDK? Normally it has a create if not exists

Reply:
It creates db & collection only if you call `CreateIfNotExists`. This method return Created object.
In my opinion it's clearer (in terms of code) to Create objects in you code in different method.

Reply:
I'm not sure I'm following I think because in that method you do what you should do here.

I think we need a GetOrCreateDatabase/GetOrCreateContainer that does this without the SemaphoreSlim